File: java/server/src/net/kio/its/server/ServerWorker.java
package net.kio.its.server;


import net.kio.its.event.Event;
import net.kio.its.logger.LogType;
import net.kio.its.responsesystem.IResponseWorker;
import net.kio.its.responsesystem.Response;
import net.kio.its.responsesystem.ResponseManager;
import net.kio.its.server.events.*;
import net.kio.security.dataencryption.EncryptedRequestManager;
import net.kio.security.dataencryption.KeysGenerator;

import java.io.*;
import java.net.Socket;
import java.net.SocketException;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

public class ServerWorker extends Thread implements IResponseWorker {

    private final Server server;
    private final Socket socket;
    private boolean isConnected;
    private boolean connectionProtocol;

    private final DataOutputStream dataOutputStream;
    private final BufferedReader bufferedReader;

    private final KeysGenerator keysGenerator;
    private final ResponseManager responseManager;
    private final String requestSeparator;
    private String macAddress;
    private final String commandPrefix;

    public ServerWorker(Server server, Socket socket, KeysGenerator keysGenerator) throws IOException {
        KeysGenerator keysGenerator1;
        this.server = server;
        this.socket = socket;
        socket.setSoTimeout(0);
        this.dataOutputStream = new DataOutputStream(socket.getOutputStream());
        this.bufferedReader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        try {
            keysGenerator1 = keysGenerator.clone();
        } catch (CloneNotSupportedException e) {
            keysGenerator1 = new KeysGenerator();
        }
        this.keysGenerator = keysGenerator1;
        this.responseManager = new ResponseManager();
        this.requestSeparator = Server.generateRequestSeparator();
        this.commandPrefix = "!";
        setName("Server Worker #" + getId());
    }

    public void stopWorker() {
        if (isConnected) {
            sendData("disconnect", (!connectionProtocol));
            interrupt();
        }
    }

    public void disconnectSocket() throws IOException {
        if (socket == null) throw new SocketException("Socket isn't connected");
        else {
            stopWorker();
            socket.close();
            isConnected = false;
            interrupt();
        }
    }

    @Override
    public void run() {
        handleClientSocket();
    }

    public void handleClientSocket() {
        try {
            startConnectionProtocol();
            String line;
            while (!Thread.interrupted() && (line = bufferedReader.readLine()) != null) {
                onLineRead(line);
            }
            disconnectSocket();
        } catch (SocketException e) {
            server.getLogger().log(LogType.CRITICAL, getName() + " - Unlegit disconnection from client " + macAddress);
            ClientSocketClosedEvent event = new ClientSocketClosedEvent(socket, this);
            server.getEventsManager().callEvent(event);
        } catch (IOException exception) {
            exception.printStackTrace();
        }
    }

    private void onLineRead(String line) {
        String decryptedData = null;
        if (keysGenerator.getSecretKey() != null && (decryptedData = EncryptedRequestManager.decrypt(line, keysGenerator)) != null) {
            onDataReceived(new String[]{decryptedData, line}, true);
        } else {
            if (keysGenerator.getSecretKey() != null)
                System.out.println(Arrays.toString(keysGenerator.getSecretKey().getEncoded()));
            onDataReceived(new String[]{line}, false);
        }
    }

    private void onDataReceived(String[] data, boolean encrypted) {

        UUID msgUUID;

        if (data[0].startsWith("response:")) {
            msgUUID = UUID.fromString(data[0].substring(9, 45));
            data[0] = data[0].substring(45);
            server.getLogger().log(LogType.DEBUG, getName() + " - Response received for UUID " + msgUUID);
            responseManager.onResponseReceived(msgUUID, data[0]);
        } else {
            msgUUID = UUID.fromString(data[0].substring(0, 36));
            data[0] = data[0].substring(36);
        }

        if (data[0].startsWith(commandPrefix)) {
            onCommandReceived(data[0], encrypted, msgUUID);
        } else {
            server.getLogger().log(LogType.DEBUG, getName() + " - Data received (encrypted=" + encrypted + ", uuid=" + msgUUID + ") : " + data[0]);
            Event event = new RawDataReceivedEvent(this, data[0], msgUUID);
            if (encrypted) {
                event = new EncryptedDataReceivedEvent(this, data[1], data[0], msgUUID);
            }
            server.getEventsManager().callEvent(event);

            if (connectionProtocol) {
                if (!onConnectionProtocolDataReceived(data[0])) {
                    server.getLogger().log(LogType.CRITICAL, "Unable to establish a connection with the client");
                    stopWorker();
                }
            }
        }

    }

    private void onCommandReceived(String command, boolean encrypted, UUID msgUUID) {
        server.getLogger().log(LogType.DEBUG, getName() + " -  Command Received (encrypted=" + encrypted + ", uuid=" + msgUUID + ") : " + command);
        CommandReceivedEvent event = new CommandReceivedEvent(this, command, commandPrefix, requestSeparator, msgUUID);
        if (event.getCommand().equalsIgnoreCase("disconnect")) {
            server.getEventsManager().callEvent(new ClientDisconnectEvent(this));
            Thread.currentThread().interrupt();
        } else {
            server.getEventsManager().callEvent(event);
        }
    }

    private void startConnectionProtocol() {
        connectionProtocol = true;
    }

    private boolean onConnectionProtocolDataReceived(String data) {
        if (data.startsWith("version:")) {
            try {
                List<Integer> clientVersion = Arrays.stream(data.replace("version:", "").replace(".", "!").split("!")).map(Integer::parseInt).collect(Collectors.toList());
                List<Integer> serverVersion = Arrays.stream(Server.getVersion().replace(".", "!").split("!")).map(Integer::parseInt).collect(Collectors.toList());
                boolean b = clientVersion.get(0).equals(serverVersion.get(0)) && clientVersion.get(1).equals(serverVersion.get(1));
                sendData("version:" + b, false);
                if (!b) {
                    return false;
                }
            } catch (NumberFormatException ignored) {
            }
            sendData("commandPrefix:" + commandPrefix, false);
            sendData("separator:" + requestSeparator, false);
        } else if (data.startsWith("macAddress:")) {
            macAddress = data.replace("macAddress:", "").toUpperCase();
            if (keysGenerator.getPublicKey() == null) keysGenerator.generateKeys(false, true);
            sendData("publicKey:" + keysGenerator.getStringPublicKey(), false);
        } else if (data.startsWith("secretKey:")) {
            EncryptedRequestManager.decryptSecretKey(data.replace("secretKey:", ""), keysGenerator);
            boolean connected = keysGenerator.getSecretKey() != null;
            sendData("connected:" + connected, false);
            connectionProtocol = !connected;
            if (connected) {
                ConnectionProtocolSuccessEvent event = new ConnectionProtocolSuccessEvent(this);
                server.getEventsManager().callEvent(event);
            }
            return connected;
        } else {
            return false;
        }
        return true;
    }


    public Response sendData(String data) {
        return sendData(data, true);
    }

    public Response sendData(String data, boolean encrypt) {
        return sendData(data, (UUID) null, encrypt);
    }

    public Response sendData(String data, String responseUUID) {
        return sendData(data, UUID.fromString(responseUUID), true);
    }

    public Response sendData(String data, String responseUUID, boolean encrypt) {
        return sendData(data, UUID.fromString(responseUUID), encrypt);
    }

    public Response sendData(String data, UUID responseUUID, boolean encrypt) {
        String rawData = data;
        Response response = new Response(this, responseUUID);
        OutputStreamWriter outputStreamWriter = new OutputStreamWriter(dataOutputStream);
        PrintWriter printWriter = new PrintWriter(outputStreamWriter, true);
        data = (responseUUID != null ? "response:" : "") + response.getMsgUUID() + data;
        if (encrypt) {
            data = EncryptedRequestManager.encrypt(data, keysGenerator);
        }
        printWriter.println(data);
        server.getLogger().log(LogType.DEBUG, getName() + " - Data sent (encrypted=" + encrypt + ", uuid=" + response.getMsgUUID() + ") : " + rawData);
        return response;
    }

    public Response sendCommand(String command, String... args) {
        return sendData(commandPrefix + command + requestSeparator + String.join(requestSeparator, args));
    }

    public String getClientMacAddress() {
        return macAddress;
    }

    public Socket getClientSocket() {
        return socket;
    }

    public boolean isConnected() {
        return isConnected;
    }

    public ResponseManager getResponseManager() {
        return responseManager;
    }
}

File: golang/server/server_worker.go
package server

import (
	"fmt"
	"net"
	"strconv"
	"strings"

	"github.com/TorkhanNetwork/Networking/golang/data_encryption"
	"github.com/TorkhanNetwork/Networking/golang/events_system"
	"github.com/TorkhanNetwork/Networking/golang/response_system"
	"github.com/google/uuid"
	"github.com/kataras/golog"
)

const INPUT_BUFFER_SIZE = 65536

type ServerWorker struct {
	Id                                           int
	server                                       *Server
	commandPrefix, requestSeparator, MacAddress  string
	connection                                   *net.TCPConn
	responseManager                              response_system.ResponseManager
	connected, authenticated, connectionProtocol bool
	quitChan                                     chan bool
}

func NewServerWorker(id int, server *Server, connection *net.TCPConn) ServerWorker {
	return ServerWorker{
		Id:               id,
		server:           server,
		commandPrefix:    "!",
		requestSeparator: server.GenerateRequestSeparator(),
		connection:       connection,
		responseManager:  response_system.NewResponseManager(),
	}
}

func (serverWorker ServerWorker) GetName() string {
	return fmt.Sprintf("  |  %s  |  ServerWorker #%d ", serverWorker.server.Name, serverWorker.Id)
}

func (serverWorker *ServerWorker) GetResponseManager() response_system.ResponseManager {
	return serverWorker.responseManager
}

func (serverWorker *ServerWorker) StartWorker() {
	serverWorker.quitChan = make(chan bool)
	go serverWorker.handleClientSocket()
}

func (serverWorker *ServerWorker) StopWorker() {
	if serverWorker.connected {
		serverWorker.SendData("disconnect", uuid.Nil, serverWorker.authenticated)
		serverWorker.quitChan <- true
	}
}

func (serverWorker *ServerWorker) DisconnectSocket() error {
	if !serverWorker.connected {
		return fmt.Errorf("%s - Socket isn't connected", serverWorker.GetName())
	}
	err := serverWorker.connection.Close()
	if err != nil {
		return err
	}
	serverWorker.StopWorker()
	serverWorker.connected = false
	serverWorker.server.RemoveServerWorker(*serverWorker)
	return nil
}

func (serverWorker *ServerWorker) handleClientSocket() {
	buf := make([]byte, INPUT_BUFFER_SIZE)
	serverWorker.startConnectionProtocol()
	for {
		mLen, err := serverWorker.connection.Read(buf)
		if err != nil {
			break
		}
		select {
		case <-serverWorker.quitChan:
			serverWorker.DisconnectSocket()
		default:
			for _, line := range strings.Split(string(buf[:mLen]), "\n") {
				serverWorker.onLineRead(line)
			}
		}
	}
	var e interface{} = NewClientDisconnectEvent(serverWorker)
	serverWorker.server.EventsManager.CallEvent((*events_system.Event)(&e))
}

func (serverWorker *ServerWorker) onLineRead(line string) {
	line = strings.TrimSpace(line)
	if line != "" {
		if decryptedData, err := data_encryption.Decrypt(line, serverWorker.server.keysGenerator); err == nil {
			serverWorker.onDataReceived([]string{decryptedData, line}, true)
		} else {
			serverWorker.onDataReceived([]string{line}, false)
		}
	}

}

func (serverWorker *ServerWorker) onDataReceived(data []string, encrypted bool) {
	var msgUUID uuid.UUID
	var err error
	if strings.HasPrefix(data[0], "response:") {
		msgUUID, err = uuid.Parse(data[0][9:45])
		if err != nil {
			golog.Errorf("%s - Unable to parse response UUID : %s", serverWorker.GetName(), err)
			return
		}
		data[0] = data[0][45:]
		golog.Debugf("%s - Response received for UUID %s", serverWorker.GetName(), msgUUID.String())
		go serverWorker.responseManager.OnResponseReceived(msgUUID, data[0])
	} else {
		var err error
		msgUUID, err = uuid.Parse(data[0][:36])
		if err != nil {
			golog.Errorf("%s - Unable to parse response UUID : %s", serverWorker.GetName(), err)
			return
		}
		data[0] = data[0][36:]
	}

	if serverWorker.commandPrefix != "" && strings.HasPrefix(data[0], serverWorker.commandPrefix) {
		serverWorker.onCommandReceived(data[0], msgUUID, encrypted)
	} else {
		golog.Debugf("%s - Data received (encrypted=%t, uuid=%s) : %s", serverWorker.GetName(), encrypted, msgUUID, data[0])
	}

	var e interface{}
	if encrypted {
		e = NewEncryptedDataReceivedEvent(serverWorker, data[1], data[0], msgUUID)
	} else {
		e = NewRawDataReceivedEvent(serverWorker, data[0], msgUUID)
	}
	serverWorker.server.EventsManager.CallEvent((*events_system.Event)(&e))

	if serverWorker.connectionProtocol {
		if !serverWorker.onConnectionProtocolDataReceived(data[0]) {
			golog.Errorf("%s - Unable to establish a connection with the client", serverWorker.GetName())
			serverWorker.StopWorker()
		}
	}

}

func (serverWorker *ServerWorker) onCommandReceived(command string, msgUUID uuid.UUID, encrypted bool) {
	golog.Debugf("%s - Command Received (encrypted=%t, uuid=%s) : %s", serverWorker.GetName(), encrypted, msgUUID, command)
	var event interface{} = NewCommandReceivedEvent(serverWorker, command, serverWorker.commandPrefix, serverWorker.requestSeparator, msgUUID)
	serverWorker.server.EventsManager.CallEvent((*events_system.Event)(&event))
}

func (serverWorker *ServerWorker) startConnectionProtocol() {
	serverWorker.connectionProtocol = true
}

func (serverWorker *ServerWorker) onConnectionProtocolDataReceived(data string) bool {
	if strings.HasPrefix(data, "version:") {
		split := strings.SplitN(data[8:], ".", 3)
		correct := len(split) >= 2 && strings.Join(split[:2], ".") == strings.Join(strings.SplitN(SERVER_VERSION, ".", 3)[:2], ".")
		serverWorker.SendData("version:"+strconv.FormatBool(correct), uuid.Nil, false)
		if !correct {
			return false
		}
		serverWorker.SendData("commandPrefix:"+serverWorker.commandPrefix, uuid.Nil, false)
		serverWorker.SendData("separator:"+serverWorker.requestSeparator, uuid.Nil, false)
	} else if strings.HasPrefix(data, "macAddress:") {
		serverWorker.MacAddress = strings.ToUpper(data[11:])
		publicKey, err := ExportRsaPublicKeyToString(&serverWorker.server.keysGenerator.PublicKey)
		if err != nil {
			return false
		}
		serverWorker.SendData("publicKey:"+publicKey, uuid.Nil, false)
	} else if strings.HasPrefix(data, "secretKey:") {
		err := data_encryption.DecryptSecretKey(data[10:], &serverWorker.server.keysGenerator)
		connected := err == nil
		serverWorker.SendData("connected:"+strconv.FormatBool(connected), uuid.Nil, false)
		serverWorker.connectionProtocol = !connected
		if connected {
			var e interface{} = NewConnectionProtocolSuccessEvent(serverWorker)
			serverWorker.server.EventsManager.CallEvent((*events_system.Event)(&e))
		}
		return connected
	} else {
		return false
	}
	return true
}

func (serverWorker *ServerWorker) SendData(data string, responseUUID uuid.UUID, encrypt bool) *response_system.Response {
	rawData := data
	response := response_system.NewResponse(serverWorker, responseUUID)
	data = response.UUID.String() + data
	if responseUUID != uuid.Nil {
		data = "response:" + data
	}
	if encrypt {
		var err error
		data, err = data_encryption.Encrypt(data, serverWorker.server.keysGenerator)
		if err != nil {
			golog.Errorf("%s - Unable to encrypt data : %s", serverWorker.GetName(), err)
			return nil
		}
	}
	_, err := serverWorker.connection.Write([]byte(data + "\n"))
	if err != nil {
		golog.Errorf("%s - Unable to send data : %s", serverWorker.GetName(), err)
		return nil
	}
	golog.Debugf("%s - Data sent (encrypted=%t, uuid=%s) : %s", serverWorker.GetName(), encrypt, response.UUID, rawData)
	return &response
}

func (serverWorker *ServerWorker) SendCommand(command string, args ...string) *response_system.Response {
	return serverWorker.SendData(serverWorker.commandPrefix+command+serverWorker.requestSeparator+strings.Join(args, serverWorker.requestSeparator), uuid.Nil, true)
}
